rust/src/index/vector/diskann.rs

```diff
@@ -0,0 +1,156 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+///! DiskANN: Fast Accurate Billion-point Nearest Neighbor Search on a Single Node
+///
+/// Modified from diskann paper. The vector store is backed by the `lance` dataset.
+mod builder;
+mod row_vertex;
+mod search;
+
+use arrow_array::UInt8Array;
+use byteorder::{ByteOrder, LE};
+
+use super::{
+    graph::{Vertex, VertexSerDe},
+    MetricType,
+};
+use crate::index::vector::pq::PQBuildParams;
+use crate::Result;
+
+pub struct DiskANNParams {
+    /// out-degree bound (R)
+    pub r: usize,
+
+    /// Distance threshold
+    pub alpha: f32,
+
+    /// Search list size
+    pub l: usize,
+
+    /// Parameters to build PQ index.
+    pub pq_params: PQBuildParams,
+
+    /// Metric type.
+    pub metric_type: MetricType,
+}
+
+// Default values from DiskANN paper.
+impl Default for DiskANNParams {
+    fn default() -> Self {
+        Self {
+            r: 90,
+            alpha: 1.2,
+            l: 100,
+            pq_params: PQBuildParams::default(),
+            metric_type: MetricType::L2,
+        }
+    }
+}
+
+impl DiskANNParams {
+    pub fn new(r: usize, alpha: f32, l: usize) -> Self {
+        Self {
+            r,
+            alpha,
+            l,
+            pq_params: PQBuildParams::default(),
+            metric_type: MetricType::L2,
+        }
+    }
+
+    pub fn r(&mut self, r: usize) -> &mut Self {
+        self.r = r;
+        self
+    }
+
+    pub fn alpha(&mut self, alpha: f32) -> &mut Self {
+        self.alpha = alpha;
+        self
+    }
+
+    pub fn l(&mut self, l: usize) -> &mut Self {
+        self.l = l;
+        self
+    }
+
+    /// Set the number of bits for PQ.
+    pub fn pq_num_bits(&mut self, nbits: usize) -> &mut Self {
+        self.pq_params.num_bits = nbits;
+        self
+    }
+
+    pub fn pq_num_sub_vectors(&mut self, m: usize) -> &mut Self {
+        self.pq_params.num_sub_vectors = m;
+        self
+    }
+
+    pub fn use_opq(&mut self, use_opq: bool) -> &mut Self {
+        self.pq_params.use_opq = use_opq;
+        self
+    }
+
+    pub fn metric_type(&mut self, metric_type: MetricType) -> &mut Self {
+        self.metric_type = metric_type;
+        self
+    }
+}
+
+/// DiskANN Vertex with PQ as persisted data.
+#[derive(Clone, Debug)]
+pub(crate) struct PQVertex {
+    row_id: u64,
+    /// PQ code for the vector.
+    pq: UInt8Array,
+}
+
+impl Vertex for PQVertex {}
+
+struct PQVertexSerDe {
+    num_sub_vectors: usize,
+
+    /// Number of bits for each PQ code.
+    num_bits: usize,
+}
+
+impl PQVertexSerDe {
+    pub(crate) fn new(num_sub_vectors: usize, num_bits: usize) -> Self {
+        Self {
+            num_sub_vectors,
+            num_bits,
+        }
+    }
+}
+
+impl VertexSerDe<PQVertex> for PQVertexSerDe {
+    #[inline]
+    fn size(&self) -> usize {
+        8 /* row_id*/ + 8 /* Length of pq */ + self.num_sub_vectors * 1 /* only 8 bits now */
```

Comment:
1. `only 8 bits now` => isn't it parameterized above as pq_num_bits?
2. what's "length of pq" (8) ?

Reply:
> only 8 bits now => isn't it parameterized above as pq_num_bits?

we have not finalized how to materialize pq bits yet. i.e., 4bits and 12 bits, whether it takes ceiling or compress a few pq into the shared bytes.

> what's "length of pq" 

Length of the PQ code in the vector, i.e., sub_vectors.